File: go/board.go
package main

import (
	"fmt"
	"strconv"
)

// 1-indexed
type Pos struct {
    row int
    col int
}

type Direction int

const (
    Right Direction = iota
    Down
    Left
    Up
)

func (d *Direction) to_string() string {
    switch *d {
    case Right:
        return "Right"
    case Down:
        return "Down"
    case Left:
        return "Left"
    case Up:
        return "Up"
    }
    return "Error"
}


type Move struct {
    direction Direction
    pos Pos
}


type Board struct {
    pieces [7][7]bool
    nb_pieces int
}


func New_board() Board {
    var b = Board {
        pieces: [7][7]bool{
            {false, false, true, true, true, false, false},
            {false, false, true, true, true, false, false},
            {true, true, true, true, true, true, true},
            {true, true, true, false, true, true, true},
            {true, true, true, true, true, true, true},
            {false, false, true, true, true, false, false},
            {false, false, true, true, true, false, false},
        },
        nb_pieces: 32,
    }

    return b;
}


func (b *Board) is_finished() bool {
    return b.nb_pieces == 1;
}

func is_position_valid(row int, col int) bool {
    return (row >= 3 && row <= 5 && col > 0 && col < 8) || (col >= 3 && col <= 5 && row > 0 && row < 8);
}

func (b *Board) contains_piece(row int, col int) bool {
    return b.pieces[row-1][col-1];
}


func (b *Board) Get_possible_moves() []Move {
    var res = make([]Move, 0);

    for row := 1; row < 8; row++ {
        for col := 1; col < 8; col++ {
            if !b.contains_piece(row, col) {
                continue;
            }
            if is_position_valid(row, col+2) && !b.contains_piece(row, col+2) && b.contains_piece(row, col+1) {
                res = append(res, Move{direction: Right, pos: Pos{row: row, col: col}});
            }
            if is_position_valid(row, col-2) && !b.contains_piece(row, col-2) && b.contains_piece(row, col-1) {
                res = append(res, Move{direction: Left, pos: Pos{row: row, col: col}});
            }
            if is_position_valid(row+2, col) && !b.contains_piece(row+2, col) && b.contains_piece(row+1, col) {
                res = append(res, Move{direction: Down, pos: Pos{row: row, col: col}});
            }
            if is_position_valid(row-2, col) && !b.contains_piece(row-2, col) && b.contains_piece(row-1, col) {
                res = append(res, Move{direction: Up, pos: Pos{row: row, col: col}});
            }
        }    
    }

    return res;
}

func (b *Board) apply_move(m Move) {
    var row = m.pos.row-1;
    var col = m.pos.col-1;

    b.nb_pieces--;

    switch m.direction {
    case Right:
        b.pieces[row][col] = false;
        b.pieces[row][col+1] = false;
        b.pieces[row][col+2] = true;
    case Down:
        b.pieces[row][col] = false;
        b.pieces[row+1][col] = false;
        b.pieces[row+2][col] = true;
    case Left:
        b.pieces[row][col] = false;
        b.pieces[row][col-1] = false;
        b.pieces[row][col-2] = true;
    case Up:
        b.pieces[row][col] = false;
        b.pieces[row-1][col] = false;
        b.pieces[row-2][col] = true;
    }
    
}

func (b *Board) undo_move(m Move) {
    var row = m.pos.row-1;
    var col = m.pos.col-1;

    b.nb_pieces++;

    switch m.direction {
    case Right:
        b.pieces[row][col] = true;
        b.pieces[row][col+1] = true;
        b.pieces[row][col+2] = false;
    case Down:
        b.pieces[row][col] = true;
        b.pieces[row+1][col] = true;
        b.pieces[row+2][col] = false;
    case Left:
        b.pieces[row][col] = true;
        b.pieces[row][col-1] = true;
        b.pieces[row][col-2] = false;
    case Up:
        b.pieces[row][col] = true;
        b.pieces[row-1][col] = true;
        b.pieces[row-2][col] = false;
    }
}


func (b *Board) to_print_string() string {
    var res = "";
    for row := 0; row < 7; row++ {
        res += strconv.Itoa(row+1) + " ";
        for col := 0; col < 6; col++ {
            if b.pieces[row][col] {
                res += "+ ";
            } else {
                res += "  ";
            }
        }

        if b.pieces[row][6] {
            res += "+\n";
        } else {
            res += " \n";
        }    
    }
    res += " ";
    for i := 1; i < 8; i++ {
        res += " " + strconv.Itoa(i);
    }
    return res;
}


func print_steps(b *Board, steps *[]Move) {
    fmt.Println(b.to_print_string())
    for i := 0; i < len(*steps); i++ {
        b.apply_move((*steps)[i]);
        fmt.Println((*steps)[i].direction.to_string(), ": (", (*steps)[i].pos.row, ",", (*steps)[i].pos.col, ") -------------------------");
        fmt.Println(b.to_print_string())
    }
}

File: rust/src/board.rs
use std::fmt::Display;


// const POSSIBLE_SOLUTION: [Move; 31] = [Move::Down(Piece(2, 4)), Move::Right(Piece(3, 2)), Move::Down(Piece(1, 3)), Move::Left(Piece(1, 5)), Move::Left(Piece(3, 4)), Move::Right(Piece(3, 1)), Move::Up(Piece(3, 5)), Move::Left(Piece(3, 7)), Move::Up(Piece(4, 3)), Move::Down(Piece(1, 3)), Move::Right(Piece(4, 1)), Move::Up(Piece(4, 3)), Move::Left(Piece(4, 5)), Move::Left(Piece(4, 7)), Move::Up(Piece(4, 5)), Move::Down(Piece(1, 5)), Move::Up(Piece(5, 3)), Move::Down(Piece(2, 3)), Move::Right(Piece(5, 1)), Move::Left(Piece(5, 4)), Move::Up(Piece(6, 5)), Move::Left(Piece(5, 7)), Move::Up(Piece(7, 3)), Move::Down(Piece(4, 3)), Move::Left(Piece(7, 5)), Move::Up(Piece(7, 3)), Move::Right(Piece(5, 2)), Move::Right(Piece(5, 4)), Move::Down(Piece(3, 5)), Move::Left(Piece(5, 6)), Move::Down(Piece(5, 4))];

// 1-indexed
#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Piece(pub u32, pub u32);


#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Direction {
    Right,
    Left,
    Up,
    Down,
}


#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Move {
    piece: Piece,
    direction: Direction,
}

impl Move {
    fn new(direction: Direction, piece: Piece) -> Move {
        Move {
            direction: direction,
            piece: piece
        }
    }
}


// #[derive(Debug, PartialEq, Default)]
// pub struct Symmetry {
//     horizontal: bool,
//     vertical: bool,
//     left_diagonal: bool,
//     right_diagonal: bool,
// }


#[derive(Clone, Copy)]
pub struct Board {
    board_size: u32,
    pieces: [[bool; 7]; 7],
    nb_pieces: u32
}


impl Board {

    pub fn new() -> Self {
        let pieces = [
            [false, false, true, true, true, false, false],
            [false, false, true, true, true, false, false],
            [true, true, true, true, true, true, true],
            [true, true, true, false, true, true, true],
            [true, true, true, true, true, true, true],
            [false, false, true, true, true, false, false],
            [false, false, true, true, true, false, false],
        ];

        let b = Board {
            board_size: 7, 
            pieces: pieces, 
            nb_pieces: 32
        };

        b
    }


    // fn get_symmetries(&self) -> Symmetry {
    //     let mut res = Symmetry::default();


    //     // row wise
    //     res.horizontal = true;
    //     for row in 0..self.pieces.len()/2 {
    //         if self.pieces[row] != self.pieces[self.pieces.len() - 1 - row] {
    //             res.horizontal = false;
    //             break;
    //         }
    //     }

    //     // column wise
    //     res.vertical = true;
    //     for col in 0..self.pieces[0].len()/2 {
    //         for row in 0..self.pieces.len() {
    //             if self.pieces[row][col] != self.pieces[row][self.pieces[0].len() - 1 - col] {
    //                 res.vertical = false;
    //             }
    //         }
    //     }

    //     // left diagonal
    //     res.left_diagonal = true;
    //     for i in 0..self.pieces.len() {
    //         for j in 0..=i {
    //             if self.pieces[i-j][i] != self.pieces[i][i-j] {
    //                 res.left_diagonal = false;
    //             }
    //         }
    //     }

    //     // right diagonal
    //     res.right_diagonal = true;
    //     for i in 0..self.pieces.len() {
    //         for j in 0..=i {
    //             if self.pieces[i-j][self.pieces[0].len() - 1 - i] != self.pieces[i][self.pieces[0].len() - 1 - (i-j)] {
    //                 res.right_diagonal = false;
    //             }
    //         }
    //     }

    //     res
    // }

    pub fn is_finished(&self) -> bool {
        self.nb_pieces == 1
    }

    // inside the cross positions
    fn position_is_valid(&self, row: u32, col: u32) -> bool {
        return (row >= 3 && row <= 5 && col > 0 && col < 8) || (col >= 3 && col <= 5 && row > 0 && row < 8);
    }

    fn contains_piece(&self, row: u32 , col: u32) -> bool {
        self.pieces[row as usize - 1][col as usize - 1]
    }


    pub fn get_possible_moves(&self) -> Vec<Move> {
        let mut moves = Vec::new();
        
        for row in 1..=self.board_size {
            for col in 1..=self.board_size {
                if !self.contains_piece(row, col) {
                    continue;
                }
                if self.position_is_valid(row, col+2) && self.contains_piece(row, col + 1) && !self.contains_piece(row, col + 2) {
                    moves.push(Move::new(Direction::Right, Piece(row, col)));
                } else if col >= 2 && self.position_is_valid(row, col-2) && self.contains_piece(row, col - 1) && !self.contains_piece(row, col - 2) {
                    moves.push(Move::new(Direction::Left, Piece(row, col)));
                } else if self.position_is_valid(row+2, col) && self.contains_piece(row + 1, col) && !self.contains_piece(row + 2, col) {
                    moves.push(Move::new(Direction::Down, Piece(row, col)));
                } else if row >= 2 && self.position_is_valid(row-2, col) && self.contains_piece(row - 1, col) && !self.contains_piece(row - 2, col) {
                    moves.push(Move::new(Direction::Up, Piece(row, col)));
                }
            }
        }
        moves
    }


    pub fn apply_move(&mut self, m: &Move) {
        let Piece(row, col) = m.piece;
        let row = row as usize - 1;
        let col = col as usize - 1;

        self.nb_pieces -= 1;

        match m.direction {
            Direction::Right => {
                self.pieces[row][col] = false;
                self.pieces[row][col + 1] = false;
                self.pieces[row][col + 2] = true;
            }
            Direction::Down => {
                self.pieces[row][col] = false;
                self.pieces[row + 1][col] = false;
                self.pieces[row + 2][col] = true;
            }
            Direction::Left => {
                self.pieces[row][col] = false;
                self.pieces[row][col - 1] = false;
                self.pieces[row][col - 2] = true;
            }
            Direction::Up => {
                self.pieces[row][col] = false;
                self.pieces[row - 1][col] = false;
                self.pieces[row - 2][col] = true;
            }
        };
    }


    pub fn undo_move(&mut self, m: &Move) {
        let Piece(row, col) = m.piece;
        let row = row as usize - 1;
        let col = col as usize - 1;

        self.nb_pieces += 1;

        match m.direction {
            Direction::Right => {
                self.pieces[row][col] = true;
                self.pieces[row][col + 1] = true;
                self.pieces[row][col + 2] = false;
            }
            Direction::Down => {
                self.pieces[row][col] = true;
                self.pieces[row + 1][col] = true;
                self.pieces[row + 2][col] = false;
            }
            Direction::Left => {
                self.pieces[row][col] = true;
                self.pieces[row][col - 1] = true;
                self.pieces[row][col - 2] = false;
            }
            Direction::Up => {
                self.pieces[row][col] = true;
                self.pieces[row - 1][col] = true;
                self.pieces[row - 2][col] = false;
            }
        };
    }

}

impl Display for Board {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for row in 0..self.pieces.len() {
            // row except last character
            write!(f, "{} ", row+1).unwrap();
            for col in 0..self.pieces[row].len()-1 {
                if self.pieces[row][col] {
                    write!(f, "+ ").unwrap();
                } else {
                    write!(f, "  ").unwrap();
                }
            }

            // last character doesnt have a space after
            if self.pieces[row][self.pieces.len() as usize - 1] {
                write!(f, "+\n").unwrap();    
            } else {
                write!(f, " \n").unwrap();
            }
        }
        write!(f, " ").unwrap();
        for i in 1..=self.pieces[0].len() {
            write!(f, " {}", i).unwrap();
        }
        write!(f, "")
    }
}



pub fn print_steps(board: &mut Board, moves: &Vec<Move>) {
    println!("{}", board);
    for m in moves {
        board.apply_move(m);
        println!("{:?}---------------------------\n{}", m, board);
    }
}
